typed-racket-lib/typed-racket/utils/redirect-contract.rkt

```diff
@@ -0,0 +1,45 @@
+#lang racket/base
+
+(require syntax/modcollapse (for-template racket/base))
+(provide make-make-redirect-to-contract)
+
+;; This produces an identifier that expands to a local-require.  It's
+;; used to implement identifiers that are protected on export from TR
+;; with contracts, but where the contracted defintion is in the
+;; #%contract-defs submodule.
+
+;; varref: a variable reference to the typed module that has the
+;;         appropriate submodule in it
+
+;; id: the name of the export from the submodule which will be
+;;     redirected-to by the local-require
+
+;; stx: the syntax object that's the argument to the macro (that is,
+;;      the stx object that's the reference to the typed identifier in
+;;      a untyped module). The funny eta-expansion with `redirect` is
+;;      so that we can recursively invoke it when the redirected id is
+;;      used in operator position.
+
+;; This code was originally written by mflatt for the plai-typed
+;; language, and then slightly adapted for TR by samth.
+
+(define (((make-make-redirect-to-contract varref) id) stx)
+  (define (redirect stx)
+    (cond
+     [(identifier? stx)
+      (with-syntax ([mp (collapse-module-path-index
+                           (module-path-index-join
+                            '(submod "." #%contract-defs)
+                            (variable-reference->module-path-index
+                             varref)))]
+                    [i (datum->syntax id (syntax-e id) stx stx)])
+        #`(let ()
+            (local-require (only-in mp [#,(datum->syntax #'mp (syntax-e #'i)) i]))
```

Comment:
Module exports are symbols, but `require` and `local-require` bind based on identifiers. This makes sure that the name binds the reference on the next line.